plugins/MidiExport/MidiExport.cpp

```diff
@@ -23,338 +23,338 @@
-						{
-							st.pop_back();
-						}
-					}
-
-					st.push_back(*it);
-					pos = it->first;
-				}
-
-				while (!st.empty())
-				{
-					writeBBPattern(pat, nv, len, st.back().first, pos, st.back().second);
-					pos = st.back().second;
-					st.pop_back();
-				}
-
-				ProcessBBNotes(nv, pos);
-				writePatternToTrack(mtrack, nv);
-				++itr;
-			}
+			processTrack(track, trackIdx++);
+		}
+		else if (track->type() == Track::BBTrack)
+		{
+			processBbTrack(track);
 		}
-		size = mtrack.writeToBuffer(buffer);
-		midiout.writeRawData((char *)buffer, size);
 	}
+	// Iterate through instrument BB tracks
+	for (Track *track : tracksBb)
+	{
+		processTrack(track, trackIdx++, true);
+	}
+	// Write all buffered data to stream
+	m_file->writeAllToStream();
 
+	// Always returns success... for now?
 	return true;
-
 }
 
-
-
-void MidiExport::writePattern(MidiNoteVector &pat, QDomNode n,
-				int base_pitch, double base_volume, int base_time)
+void MidiExport::processTrack(Track *track, size_t trackIdx, bool isBb)
 {
-	// TODO interpret steps="12" muted="0" type="1" name="Piano1"  len="2592"
-	for (QDomNode nn = n.firstChild(); !nn.isNull(); nn = nn.nextSibling())
+	// Cast track as a instrument one and save info from it to element
+	InstrumentTrack *instTrack = dynamic_cast<InstrumentTrack *>(track);
+	QDomElement root = instTrack->saveState(m_dataFile, m_dataFile.content());
+
+	// Get next MIDI file track object of the list and set its channel
+	MidiFile::Track &midiTrack = m_file->m_tracks[trackIdx];
+	if (m_channel == 9) { ++m_channel; }
+	midiTrack.m_channel = m_channel++;
+
+	// Add info about tempo and track name
+	midiTrack.addTempo(m_tempo, 0);
+	midiTrack.addName(track->name().toStdString(), 0);
+
+	// If the current track is a Sf2 Player one, set the current
+	// patch to the exporting track. Note that this only works
+	// decently if the current bank is a GM 1~128 one (which would be
+	// needed as the default either way for successful import).
+	// BB tracks are always bank 128 (see MidiImport), patch 0.
+	uint8_t patch = 0;
+	QString instName = instTrack->instrumentName();
+	if (instName == "Sf2 Player")
 	{
-		QDomElement note = nn.toElement();
-		if (note.attribute("len", "0") == "0") continue;
-		// TODO interpret pan="0" fxch="0" pitchrange="1"
-		MidiNote mnote;
-		mnote.pitch = qMax(0, qMin(127, note.attribute("key", "0").toInt() + base_pitch));
-		 // Map from LMMS volume to MIDI velocity
-		mnote.volume = qMin(qRound(base_volume * LocaleHelper::toDouble(note.attribute("vol", "100")) * (127.0 / 200.0)), 127);
-		mnote.time = base_time + note.attribute("pos", "0").toInt();
-		mnote.duration = note.attribute("len", "0").toInt();
-		pat.push_back(mnote);
+		class Instrument *inst = instTrack->instrument();
+		uint8_t bank = inst->childModel("bank")->value<uint8_t>();
+		if (bank == 128)
+		{
+			// Drum Sf2 track, so set its channel to 10
+			// (and reverse counter increment)
+			midiTrack.m_channel = 9;
+			m_channel--;
```

Comment:
If `m_channel` was 9 before calling this function, this line will change it to 10 instead of going back to 9.